src/Input.cpp

```diff
@@ -210,63 +339,82 @@ struct Gosu::Input::Impl
             }
         }
     }
-    
+
     void dispatch_enqueued_events()
     {
-        for (int event : event_queue) {
-            bool down = (event >= 0);
-            Button button(down ? event : ~event);
-            
-            button_states[button.id()] = down;
-            if (down && input.on_button_down) {
-                input.on_button_down(button);
+        for (InputEvent event : event_queue) {
+            if (event.type == InputEventType::ButtonEvent) {
+                bool down = (event.id >= 0);
+                Button button(down ? event.id : ~event.id);
+
+                button_states[button.id()] = down;
+                if (down && input.on_button_down) {
+                    input.on_button_down(button);
+                }
+                else if (!down && input.on_button_up) {
+                    input.on_button_up(button);
+                }
             }
-            else if (!down && input.on_button_up) {
-                input.on_button_up(button);
+            else if (event.type == InputEventType::ControllerConnectionEvent) {
+                if (event.gamepad_connected) {
+                    if (input.on_gamepad_connected) {
+                        input.on_gamepad_connected(event.id);
+                    }
+                }
+                else {
+                    if (input.on_gamepad_disconnected) {
+                        input.on_gamepad_disconnected(event.id);
+                        free_gamepad_slot_index(event.gamepad_instance_id);
+                    }
+                }
             }
         }
         event_queue.clear();
     }
-    
+
 private:
     // For button down event: Button name value (>= 0)
     // For button up event: ~Button name value (< 0)
-    vector<int> event_queue;
+    vector<InputEvent> event_queue;
 
     void enqueue_event(int id, bool down)
     {
-        event_queue.push_back(down ? id : ~id);
+        InputEvent s = { down ? id : ~id, InputEventType::ButtonEvent };
```

Comment:
The silly id/~id trick was just because I was too lazy to create a new struct instead of using `int`s. Now that there's a struct, we should make this more explicit, for example by creating a new InputEventType.